File: stockpilot-app/android/app/app/src/main/java/com/stockpilot/app/core/Report.kt
package com.stockpilot.app.core

import java.util.Locale

/** 一次盯盘扫描的单只股票结果（含通知决策） */
data class WatchResult(
    val stock: StockRef,
    val signal: Signal,
    val price: Double,
    val changePct: Double,
    val mainNet: Double?,
    val notify: Boolean,
    val notifyTitle: String,
    val notifyBody: String
)

/** 盘后汇总报告生成（对应 core/report.ts） */
object Report {

    fun fmtMoney(v: Double?): String {
        if (v == null || v.isNaN() || v.isInfinite()) return "--"
        val a = Math.abs(v)
        val sign = if (v > 0) "+" else if (v < 0) "-" else ""
        return when {
            a >= 1e8 -> sign + String.format(Locale.US, "%.2f", a / 1e8) + "亿"
            a >= 1e4 -> sign + String.format(Locale.US, "%.1f", a / 1e4) + "万"
            else -> sign + String.format(Locale.US, "%.0f", a)
        }
    }

    fun fmtPct(v: Double): String {
        if (v.isNaN() || v.isInfinite()) return "--"
        return (if (v >= 0) "+" else "") + String.format(Locale.US, "%.2f", v) + "%"
    }

    fun buildDaily(date: String, results: List<WatchResult>, phase: PhaseInfo): String {
        val buy = results.filter { it.signal.action == "买入" }
        val sell = results.filter { it.signal.action == "卖出" }
        val sorted = results.sortedByDescending { it.changePct }
        val topUp = sorted.take(5)
        val topDown = sorted.takeLast(5).reversed().filter { it !in topUp }

        val sb = StringBuilder()
        sb.append("# StockPilot 盘后汇总 ").append(date).append('\n').append('\n')
        sb.append("市场阶段：").append(phase.label).append("（").append(phase.note).append("）\n")
        sb.append("自选数量：").append(results.size)
            .append("｜买入信号：").append(buy.size)
            .append("｜卖出信号：").append(sell.size).append('\n').append('\n')

        if (buy.isNotEmpty()) {
            sb.append("## 买入信号\n")
            for (r in buy) {
                sb.append("- ").append(r.stock.name).append("(").append(r.stock.code).append(") 现价 ")
                    .append(String.format(Locale.US, "%.2f", r.price)).append(' ').append(fmtPct(r.changePct))
                    .append("｜评分 ").append(r.signal.score)
                    .append("｜主力 ").append(fmtMoney(r.mainNet)).append('\n')
                sb.append("  依据：").append(r.signal.sessionSummary).append('\n')
                val tech = r.signal.reasons.filter { !it.startsWith("【") }.take(2).joinToString("；")
                if (tech.isNotEmpty()) sb.append("  技术面：").append(tech).append('\n')
            }
            sb.append('\n')
        }

        if (sell.isNotEmpty()) {
            sb.append("## 卖出信号\n")
            for (r in sell) {
                sb.append("- ").append(r.stock.name).append("(").append(r.stock.code).append(") 现价 ")
                    .append(String.format(Locale.US, "%.2f", r.price)).append(' ').append(fmtPct(r.changePct))
                    .append("｜评分 ").append(r.signal.score)
                    .append("｜主力 ").append(fmtMoney(r.mainNet)).append('\n')
                sb.append("  依据：").append(r.signal.sessionSummary).append('\n')
            }
            sb.append('\n')
        }

        sb.append("## 自选表现（涨幅前五）\n")
        for (r in topUp) {
            sb.append("- ").append(r.stock.name).append(' ').append(fmtPct(r.changePct))
                .append("｜评分 ").append(r.signal.score).append('｜').append(r.signal.action).append('\n')
        }
        if (topDown.isNotEmpty()) {
            sb.append('\n').append("## 自选表现（跌幅前五）\n")
            for (r in topDown) {
                sb.append("- ").append(r.stock.name).append(' ').append(fmtPct(r.changePct))
                    .append("｜评分 ").append(r.signal.score).append('｜').append(r.signal.action).append('\n')
            }
        }
        sb.append('\n').append("> 数据来自公开行情接口，策略为技术指标分析，仅供参考，不构成投资建议。")
        return sb.toString()
    }

    /** 用于通知栏的一行摘要 */
    fun summaryLine(results: List<WatchResult>): String {
        val buy = results.count { it.signal.action == "买入" }
        val sell = results.count { it.signal.action == "卖出" }
        return "自选 ${results.size}｜买入 $buy｜卖出 $sell"
    }
}

File: stockpilot-app/harmony/entry/src/main/ets/core/report.ts
// core/report.ts
// 盘后汇总报告生成（纯函数，便于在鸿蒙端直接渲染或保存）。

import type { WatchResult, PhaseInfo } from './types';

function fmtMoney(v: number | null): string {
  if (v == null || !isFinite(v)) return '--';
  const a = Math.abs(v);
  const sign = v > 0 ? '+' : (v < 0 ? '-' : '');
  if (a >= 1e8) return sign + (a / 1e8).toFixed(2) + '亿';
  if (a >= 1e4) return sign + (a / 1e4).toFixed(1) + '万';
  return sign + a.toFixed(0);
}

function fmtPct(v: number): string {
  if (v == null || !isFinite(v)) return '--';
  return (v >= 0 ? '+' : '') + v.toFixed(2) + '%';
}

export interface DailyReport {
  date: string;
  title: string;
  text: string;          // 可在通知/分享中使用的纯文本
  buyCount: number;
  sellCount: number;
  watchCount: number;
  topUp: WatchResult[];
  topDown: WatchResult[];
}

export function buildDailyReport(date: string, results: WatchResult[], phase: PhaseInfo): DailyReport {
  const buy = results.filter(r => r.signal.action === '买入');
  const sell = results.filter(r => r.signal.action === '卖出');
  const sorted = results.slice().sort((a, b) => b.changePct - a.changePct);
  const topUp = sorted.slice(0, 5);
  const topDown = sorted.slice(-5).reverse().filter(r => topUp.indexOf(r) < 0);

  const lines: string[] = [];
  lines.push(`# StockPilot 盘后汇总 ${date}`);
  lines.push('');
  lines.push(`市场阶段：${phase.label}（${phase.note}）`);
  lines.push(`自选数量：${results.length}｜买入信号：${buy.length}｜卖出信号：${sell.length}`);
  lines.push('');

  if (buy.length > 0) {
    lines.push('## 买入信号');
    for (const r of buy) {
      lines.push(`- ${r.stock.name}(${r.stock.code}) 现价 ${r.price.toFixed(2)} ${fmtPct(r.changePct)}｜评分 ${r.signal.score}｜主力 ${fmtMoney(r.mainNet)}`);
      lines.push(`  依据：${r.signal.sessionSummary}`);
      const tech = r.signal.reasons.filter(x => x.indexOf('【') !== 0).slice(0, 2).join('；');
      if (tech) lines.push(`  技术面：${tech}`);
    }
    lines.push('');
  }

  if (sell.length > 0) {
    lines.push('## 卖出信号');
    for (const r of sell) {
      lines.push(`- ${r.stock.name}(${r.stock.code}) 现价 ${r.price.toFixed(2)} ${fmtPct(r.changePct)}｜评分 ${r.signal.score}｜主力 ${fmtMoney(r.mainNet)}`);
      lines.push(`  依据：${r.signal.sessionSummary}`);
    }
    lines.push('');
  }

  lines.push('## 自选表现（涨幅前五）');
  for (const r of topUp) {
    lines.push(`- ${r.stock.name} ${fmtPct(r.changePct)}｜评分 ${r.signal.score}｜${r.signal.action}`);
  }
  lines.push('');
  lines.push('## 自选表现（跌幅前五）');
  for (const r of topDown) {
    lines.push(`- ${r.stock.name} ${fmtPct(r.changePct)}｜评分 ${r.signal.score}｜${r.signal.action}`);
  }
  lines.push('');
  lines.push('> 数据来自公开行情接口，策略为技术指标分析，仅供参考，不构成投资建议。');

  const text = lines.join('\n');
  return {
    date,
    title: `StockPilot 盘后汇总 ${date}（买${buy.length}/卖${sell.length}）`,
    text,
    buyCount: buy.length,
    sellCount: sell.length,
    watchCount: results.length,
    topUp, topDown
  };
}
